bpf/cachingmap/caching_map.go

```diff
@@ -0,0 +1,345 @@
+}
+
+func (c *CachingMap) clearCache() {
+	logrus.WithField("name", c.params.Name).Debug("Clearing cache of BPF map")
+	c.cacheOfDataplane = nil
+	c.pendingDeletions = nil
+	c.pendingUpdates = nil
+}
+
+func (c *CachingMap) recalculatePendingOperations() {
+	debug := logrus.GetLevel() >= logrus.DebugLevel
+
+	// Look for any discrepancies and queue up updates.
+	c.desiredStateOfDataplane.Iter(func(k, desiredVal []byte) {
+		actualVal := c.cacheOfDataplane.Get(k)
+		if reflect.DeepEqual(actualVal, desiredVal) {
+			return
+		}
+		c.pendingUpdates.Set(k, desiredVal)
+	})
+
+	// Scan for any dataplane keys that are not in the desired map at all and
+	// queue up deletions.
+	c.cacheOfDataplane.Iter(func(k, actualVal []byte) {
+		desiredVal := c.desiredStateOfDataplane.Get(k)
+		if debug {
+			logrus.WithFields(logrus.Fields{
+				"k":        k,
+				"v":        actualVal,
+				"expected": desiredVal,
+			}).Debug("Checking cache against desired")
+		}
+		if desiredVal == nil {
+			c.pendingDeletions.Set(k, actualVal)
+			return
+		}
+	})
+
+	logrus.WithFields(logrus.Fields{
+		"cached":         c.cacheOfDataplane.Len(),
+		"pendingDels":    c.pendingDeletions.Len(),
+		"pendingUpdates": c.pendingUpdates.Len(),
+		"name":           c.params.Name,
+	}).Info("Recalculated pending operations")
+}
+
+func (c *CachingMap) SetDesiredState(k, v []byte) {
+	if logrus.GetLevel() >= logrus.DebugLevel {
+		logrus.WithFields(logrus.Fields{"name": c.params.Name, "k": k, "v": v}).Debug("SetDesiredState")
+	}
+	c.desiredStateOfDataplane.Set(k, v)
+	if c.cacheOfDataplane == nil {
+		logrus.Debug("SetDesiredState: initial sync pending.")
+		return // Initial sync is pending, we're not tracking deltas yet.
+	}
+	c.pendingDeletions.Delete(k)
+
+	// Check if we think we need to update the dataplane as a result.
+	currentVal := c.cacheOfDataplane.Get(k)
+	if reflect.DeepEqual(currentVal, v) {
+		// Dataplane already agrees with the new value so clear any pending update.
+		logrus.Debug("SetDesiredState: Key in dataplane already, ignoring.")
+		c.pendingUpdates.Delete(k)
+		return
+	}
+	c.pendingUpdates.Set(k, v)
+}
+
+func (c *CachingMap) DeleteDesiredState(k []byte) {
+	if logrus.GetLevel() >= logrus.DebugLevel {
+		logrus.WithFields(logrus.Fields{"name": c.params.Name, "k": k}).Debug("DeleteDesiredState")
+	}
+	c.desiredStateOfDataplane.Delete(k)
+	if c.cacheOfDataplane == nil {
+		logrus.Debug("DeleteDesiredState: initial sync pending.")
+		return // Initial sync is pending, we're not tracking deltas yet.
+	}
+	c.pendingUpdates.Delete(k)
+
+	// Check if we need to update the dataplane.
+	currentVal := c.cacheOfDataplane.Get(k)
+	if currentVal == nil {
+		// We don't think this value is in the dataplane so clear any pending delete.
+		logrus.Debug("DeleteDesiredState: Key not in dataplane, ignoring.")
+		c.pendingDeletions.Delete(k)
+		return
+	}
+	c.pendingDeletions.Set(k, currentVal)
+}
+
+func (c *CachingMap) DeleteAll() {
```

Comment:
nit: the DeleteAll() name is a bit confusing. Could it be something like ResetDesiredState, I am not sure.

Reply:
Adjusted the naming to be consistent and added comments.